readme.md
Adds a hotkey (default F12) that opens a small gui to take screenshots.

Default values are:  
Resolution : 2048 x 2048  
Position   : Centered on player  
Zoom       : 1.0  
Filename   : screenshot_<current-game-tick>.png  
show_entity_info : true  (shows 'alt'-info like interter arrows and assembler recipies)  
show_gui         : false (doesn't seem to do anything currently)  
anti_alias       : false (quadtruples system burden increasing likelyhood of game freeze)  

Resolution and Zoom are stored for the duration of the session.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 3.0.12
Date: 2021-08-18
  Bugfixes:
    - Fixed (yet another) crash when trying to rotate without selecting a canvas.
---------------------------------------------------------------------------------------------------
Version: 3.0.11
Date: 2021-04-26
  Internal:
    - Changed custom inputs to use erlib-style on_linked_input_* names.
  Bugfixes:
    - Removed one pixel wide white border from screenshots without size labels.
---------------------------------------------------------------------------------------------------
Version: 3.0.10
Date: 2021-04-24
  Bugfixes:
    - Size Labels option is now factored into resolution and file size estimate.
    - Resolution and file size estimate are now also shown for full-screen mode.
---------------------------------------------------------------------------------------------------
Version: 3.0.9
Date: 2021-04-16
  Bugfixes:
    - Added missing handler for on_player_removed event. (Replaces 3.0.8 hotfix.)
---------------------------------------------------------------------------------------------------
Version: 3.0.8
Date: 2021-04-15
  Bugfixes:
    - Hotfixed invalid LuaPlayer cache, but don't really know how that's possible.
---------------------------------------------------------------------------------------------------
Version: 3.0.7
Date: 2021-03-16
  Minor Features:
    - Added Japanese locale.
    - Added German locale.
  Bugfixes:
    - Some gui content frames had wrong width.
    - Changed final frame to have variable size buttons for easier localization.
---------------------------------------------------------------------------------------------------
Version: 3.0.6
Date: 2021-03-15
  Bugfixes:
    - Marked as incompatible with old mod name.
---------------------------------------------------------------------------------------------------
Version: 3.0.5
Date: 2021-03-15
  Major Features:
    - Updated factorio version to 1.1.
    - Changed internal mod name to "eradicators-screenshot-maker".
  Bugfixes:
    - Gui click sound volume was zoom dependant.
    - Fixed zoom text field had wrong size after style change.
---------------------------------------------------------------------------------------------------
Version: 3.0.4
Date: 2020-12-13
  Bugfixes:
    - Fixed crash when trying to rotate without selecting a canvas.
---------------------------------------------------------------------------------------------------
Version: 3.0.3
Date: 2020-12-13
  Minor Features:
    - Added option to include canvas frame size in screenshot.
    - Added Jpeg quality support.
    - Added canvas frame rotation support.
    - Added tooltips with extensive usage instructions.
    - Added partial gui hiding in screenshot mode.
    - Added ability to make full-screen screenshots when no canvas is selected.
    - Added day or night illumination support.
---------------------------------------------------------------------------------------------------
Version: 3.0.2
Date: 2020-12-10
  Known Issues:
    - [Factorio 1.0 Engine Bug] After loading a save you have to re-open the gui or the option toggles won't work.
---------------------------------------------------------------------------------------------------
Version: 3.0.0
Date: 2020-12-03
  Major Features:
    - Rewritten from scratch to support modern engine features.
    - Screenshot area now shows a live preview while dragging.
    - Screenshot area can now be resized and moved after initial selection.
    - Screenshot resolution and approximate size are now shown.
    - Added support for different file formats.
    - The GUI is now always shown while you hold the camera.
    - Screenshot mode can now also be activated via quickbar shortcut, and the default hotkey was changed to "PrintScreen".
  Removed Features:
    - Anti-Aliasing support was removed because it used a non-moddable low-quality in-engine rescaler. I recommend using an external image program for high-quality downsampling instead.
---------------------------------------------------------------------------------------------------
Version: 2.1.3
Date: 2020-12-03
  Major Features:
    - Updated factorio version to 1.0.
---------------------------------------------------------------------------------------------------
Version: 2.1.1
Date: 2019-08-03
  Bugfixes:
    - Manual zoom override does not change the size of the captured area anymore.
    - Added in-game mod menu thumbnail.
---------------------------------------------------------------------------------------------------
Version: 2.1.0
Date: 2019-06-09
  Major Features:
    - Updated factorio version to 0.17.
  Internal:
    - Camera is now a cursor-only item.
---------------------------------------------------------------------------------------------------
Version: 2.0.0
Date: 2018-07-09
  Bugfixes:
    - Fixed possible multiplayer desync.
    - Renamed "zoom" to "scale" internally to avoid name conflict with LuaGuiElement method
    - Changed version name scheming to be independant of factorio version.
    - Added range check so invalid resolutions don't cause errors.
  Major Features:
    - Added ability to draw a camera box around the target area instead of specifying coordinates.
    - Added ability to hold shift while using the camera tool to skip the GUI and take a screenshot directly.
    - Added a camera shutter confirmation sound.
    - Added tile-perfect screenshots per default. Non-perfect screenshots are still possible using the GUI.
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2016-07-30
  Major Features:
    - Initial Release.


data.lua
--TODO: make invisible cursor box (copy "make_cursor_box" function to make util sprites).

local CONST = require 'control/constants'


local asset = function(path) return (CONST.MOD_ROOT..'assets/'..path):gsub('/+','/') end


local ICONS_96 = {{
  icon     = asset 'camera.png',
  icon_size = 96,
  }}
  
local SPRITE_96 = {
  filename = asset 'camera.png',
  size = 96,
  }

-- -------------------------------------------------------------------------- --
-- Hotkey                                                                     --
-- -------------------------------------------------------------------------- --

data:extend{{ type = 'custom-input',
  name = CONST.ITEM_NAME,
  key_sequence = 'PRINTSCREEN',
  -- key_sequence = 'F12',
  order = 'a',
  localised_name = {'shortcut-name.er:screenshot-camera'},
  }}

-- -------------------------------------------------------------------------- --
-- Shortcut                                                                   --
-- -------------------------------------------------------------------------- --

data:extend{{
  type = 'shortcut',
  name = CONST.ITEM_NAME,
  -- action = 'create-blueprint-item',
  action = 'lua',
  -- item_to_create = CONST.ITEM_NAME,
  associated_control_input = CONST.ITEM_NAME,
  icon = SPRITE_96,
  toggleable = true,
  }}

-- -------------------------------------------------------------------------- --
-- Sound                                                                      --
-- -------------------------------------------------------------------------- --

data:extend{{
  type = "sound",
  name = 'er:camera-click',
  variations = {
    {filename = asset '/camera-click.ogg'},
    },
  volume = 1
	}}

data:extend{{
  type = "sound",
  name = 'er:gui-switch-click',
  variations = {{
    filename = "__core__/sound/gui-switch.ogg",
    -- volume = 1,
    -- preload = false,
    }},
  -- filename = "__core__/sound/gui-switch.ogg",
  -- volume = 1
  audible_distance_modifier = 1,
	}}
  
  
  
-- -------------------------------------------------------------------------- --
-- Item                                                                       --
-- -------------------------------------------------------------------------- --

  
-- for _,control in pairs{
--   -- 'rotate',
--   -- 'build','mine',
--   -- 'open-gui',
--   -- 'move-down','move-left','move-right','move-up',
--   'clean-cursor',
--   } do
--   local name = 'er:controls:'..control
--   if not (data.raw['custom-input'] and data.raw['custom-input'][name]) then
--     data:extend{{
--       type                = 'custom-input' ,
--       name                = name           ,
--       key_sequence        = ''             , --nil is invalid
--       linked_game_control = control        }}
--     end
--   end


--[[
data:extend{{
  -- This allows loading the selection-tool type item when mods are removed
  type = 'selection-tool',
  name = CONST.ITEM_NAME .. '-2',
  -- localised_name = {'item-name.blueprint'},
  icon = asset '/camera.png',
  icon_size = 96,
  -- flags = {'goes-to-quickbar', 'hidden'}, --0.16
  flags = {'hidden','only-in-cursor'}, --0.17
  -- subgroup = 'tool',
  -- order = 'c[automated-construction]-a[blueprint]',
  stack_size = 1,
  stackable = false,
  selection_color = { r = 0, g = 0, b = 0 },
  alt_selection_color = { r = 1, g = 1, b = 1 , a = 0}, --alt mode invisible == disabled
  -- Valid values are: blueprint, deconstruct, cancel-deconstruct, items, trees, buildable-type, tiles, items-to-place, any-entity, any-tile, matches-force
  -- selection_mode = {'blueprint'},
  -- alt_selection_mode = {'blueprint'},
  selection_mode = {'any-entity'},
  alt_selection_mode = {'any-entity'},
  selection_cursor_box_type = 'copy',
  alt_selection_cursor_box_type = 'copy',
  show_in_library = false
  }}
--]]



data:extend{{
  type = 'capsule',
  name = CONST.ITEM_NAME,
  -- flags = {"hidden", "only-in-cursor", "spawnable"},
  flags = {"hidden", "only-in-cursor","not-stackable"}, --not stackable flag removes stack size number! :D
  -- draw_label_for_cursor_render = false,
  mouse_cursor = "selection-tool-cursor",
  icons = ICONS_96,
  radius_color = {a=0}, -- don't show the radius
  stack_size = 1, -- abuse stacksize to show selection size?
  stackable = false,
  capsule_action = {
    type = 'throw',
    uses_stack = false,
    attack_parameters = {
      type = 'projectile',
      range = 1e7,
      cooldown = 2, -- 33 milliseconds
      ammo_type = {
        category = 'melee',
        }
      },
    }
  }}

  
--[[
-- Artillery shows artillery-turrents-in-range count and errors if that's 0
  
data:extend{{
  type = 'capsule',
  name = CONST.ITEM_NAME,
  -- flags = {"hidden", "only-in-cursor", "spawnable"},
  flags = {"hidden", "only-in-cursor","not-stackable"}, --not stackable flag removes stack size number! :D
  -- draw_label_for_cursor_render = false,
  mouse_cursor = "selection-tool-cursor",
  icons = ICONS_96,
  radius_color = {a=0}, -- don't show the radius
  stack_size = 1, -- abuse stacksize to show selection size?
  stackable = false,
  capsule_action = {
    type = "artillery-remote", -- artillery type can be used from map
    flare = "artillery-flare-2"
    },
  }}

data:extend{{
    type = "artillery-flare",
    name = "artillery-flare-2",
    icon = "__base__/graphics/icons/artillery-targeting-remote.png",
    icon_size = 64, icon_mipmaps = 4,
    flags = {"placeable-off-grid", "not-on-map"},
    map_color = {r=1, g=0.5, b=0},
    -- life_time = 60 * 60,
    life_time = 0,
    initial_height = 0,
    initial_vertical_speed = 0,
    initial_frame_speed = 1,
    -- shots_per_flare = 1,
    shots_per_flare = 0,
    -- early_death_ticks = 3 * 60,
    early_death_ticks = 0,
    pictures =
    {
      {
        filename = "__core__/graphics/shoot-cursor-red.png",
        priority = "low",
        width = 258,
        height = 183,
        frame_count = 1,
        scale = 1,
        flags = {"icon"}
      },
      --{
      --  filename = "__base__/graphics/entity/sparks/sparks-02.png",
      --  width = 36,
      --  height = 32,
      --  frame_count = 19,
      --  line_length = 19,
      --  shift = {0.03125, 0.125},
      --  tint = { r = 1.0, g = 0.9, b = 0.0, a = 1.0 },
      --  animation_speed = 0.3,
      --}
    }
  }}
--]]


data:extend{{
  type = "font",
  name = "er:selection-rectangle-label-font",
  from = "default-semibold",
  size = 32,
  }}
  
  
data:extend{{
  key_sequence                 = ''                                ,
  action                       = nil                               ,
  alternative_key_sequence     = nil                               ,
  consuming                    = nil                               ,
  enabled                      = true                              ,
  enabled_while_in_cutscene    = false                             ,
  enabled_while_spectating     = false                             ,
  item_to_create               = nil                               ,
  linked_game_control          = 'mine'                            ,
  name                         = CONST.EVENT.RIGHT_CLICK           ,
  type                         = 'custom-input'                    ,
  localised_description        = nil                               ,
  localised_name               = nil ,
  order                        = 'a'                               ,
  }}

data:extend{{
  key_sequence                 = ''                       ,
  enabled                      = true                     ,
  linked_game_control          = 'rotate'                 ,
  name                         = CONST.EVENT.ROTATE_RIGHT ,
  type                         = 'custom-input'           ,
  order                        = 'z1'                     ,
  }}
data:extend{{
  key_sequence                 = ''                       ,
  enabled                      = true                     ,
  linked_game_control          = 'reverse-rotate'         ,
  name                         = CONST.EVENT.ROTATE_LEFT  ,
  type                         = 'custom-input'           ,
  order                        = 'z1'                     ,
  }}
  

local styles = data.raw['gui-style'].default
  
styles['er:screenshot-gui-status-table'] = {
  type = 'table_style',
  parent = nil,
  column_widths = {
    {column = 1, width = (256 - 24) / 2}, -- 116
    {column = 2, width = (256 - 24) / 2},
    },
  column_alignments = {
    {column = 1, alignment = 'middle-left'  },
    {column = 2, alignment = 'middle-center'},
    }
  }

styles['er:screenshot-gui-advanced-option-table'] = {
  type = 'table_style',
  parent = nil,
  column_widths = {
    -- {column = 1, width = (256 - 24) / 2}, -- 116
    -- {column = 2, width = (256 - 24) / 2},
    {column = 1, width = (256 - 24) - 140}, -- 116
    {column = 2, width = 140},
    },
  column_alignments = {
    -- {column = 1, alignment = 'middle-center'},
    {column = 1, alignment = 'middle-left'},
    {column = 2, alignment = 'middle-center'},
    }
  }
  
-- -------------------------------------------------------------------------- --
-- Aligned Switch Style                                                       --
-- -------------------------------------------------------------------------- --
  
styles['er:screenshot-gui-aligned-switch-table'] = {
  type = 'table_style',
  parent = nil,
  -- width = (256 - 24) / 2,
  -- width = 
  column_widths = {
    {column = 1, width = 54}, -- 116
    -- {column = 2, width = 40}, --default switch size
    {column = 3, width = 54},
    },
  column_alignments = {
    {column = 1, alignment = 'middle-right' },
    {column = 2, alignment = 'middle-center'},
    {column = 3, alignment = 'middle-left'  },
    }
  }
-- Create stand-alone names for vanilla switch labels.
styles['er:screenshot-gui-aligned-switch-inactive-label'] = 
  util.table.deepcopy(styles['switch'].inactive_label)
styles['er:screenshot-gui-aligned-switch-active-label'] = 
  util.table.deepcopy(styles['switch'].active_label)

-- styles['er:screenshot-gui-aligned-switch-inactive-label']
  -- .left_click_sound = {{ filename = "__core__/sound/gui-switch.ogg", volume = 0.7 }}
-- styles['er:screenshot-gui-aligned-switch-active-label']
  -- .left_click_sound = {{ filename = "__core__/sound/gui-switch.ogg", volume = 0.7 }}
  
  
  

styles['er:screenshot-gui-bold-green-label'] = {
  type = "label_style",
  parent = "bold_label",
  font_color = {0, 1, 0}
  }

styles['er:screenshot-gui-bold-yellow-label'] = {
  type = "label_style",
  parent = "bold_label",
  font_color = {1, 1, 0}
  }

styles['er:screenshot-gui-bold-red-label'] = {
  type = "label_style",
  parent = "bold_label",
  font_color = {1, 0, 0}
  }
  

  
  
-- local switch_label_width = 38
local switch_label_width = 64

styles['er:screenshot-gui-advanced-option-switch'] = {
  type = 'switch_style',
  parent = 'switch',
  -- width = 116,
  horizontally_stretchable = 'off',
  horizontally_squashable = 'off',
  
  maximum_horizontal_squash_size = 116,
  
  active_label =
  {
    type = "label_style",
    font_color = {241, 190, 100},
    font = "default-bold",
    
    width = switch_label_width,
    horizontally_stretchable = 'off',
    horizontally_squashable = 'off',
    -- size = {switch_label_width,24},
    minimal_width = switch_label_width,
    maximal_width = switch_label_width,
    natural_width = switch_label_width,
    maximum_horizontal_squash_size = 116,  
    
  },
  inactive_label =
  {
    type = "label_style",
    font_color = default_font_color,
    hovered_font_color = {255, 230, 192},
    font = "default",

    width = switch_label_width,
    horizontally_stretchable = 'off',
    horizontally_squashable = 'off',
    -- size = {switch_label_width,24},
    minimal_width = switch_label_width,
    maximal_width = switch_label_width,
    natural_width = switch_label_width,
      
    maximum_horizontal_squash_size = 116,
  },
  
  flow = {},
  }